docs/upgrade/money-class.md

```diff
@@ -0,0 +1,430 @@
+- `\Tests\ShopBundle\Functional\Model\Product\ProductOnCurrentDomainFacadeTest`
+- `\Tests\ShopBundle\Functional\Model\Product\ProductVisibilityRepositoryTest`
+- `\Tests\ShopBundle\Functional\PersonalData\PersonalDataExportXmlTest`
+- `\Tests\ShopBundle\Functional\Twig\PriceExtensionTest`
+
+Your [HTTP smoke tests](/docs/introduction/automated-testing.md#http-smoke-tests) and [acceptance test](/docs/introduction/automated-testing.md#acceptance-tests-aka-functional-tests-or-selenium-tests) shouldn't require any changes.
+
+## Javascript
+The `Money` class implements the `JsonSerializable` interface so it can be serialized using `json_encode` into an object usable in JS.
+
+You can read more about it in [the Money in Javascript section](/docs/introduction/how-to-work-with-money.md#money-in-javascript) of the documentation if you work with monetary values using JS in your project.
+There were no changes needed in `shopsys/project-base`.
+
+## Other
+The method `Cart::getQuantifiedProductsIndexedByItemId()` was renamed to `Cart::getQuantifiedProducts()` and no longer uses the product IDs for indexing.
+This was done due to the fact that it could lead to unexpected behavior when the `CartItem` entities have not yet been persisted (and thus they don't have been assigned an ID).
+
+Similarly, `CartFacade::getQuantifiedProductsOfCurrentCustomerIndexedByCartItemId()` was renamed to `CartFacade::getQuantifiedProductsOfCurrentCustomer()` and also no longer uses the product IDs for indexing.
+
+Check your usage of the methods and make appropriate changes.
+
+Note that the indexing may be used in Twig templates as well.
+As an example, see the change that was required in `@ShopsysShop/Front/Content/Cart/index.html.twig` in `shopsys/project-base`:
+```diff
+
+    <tbody>
+-       {% for cartItem in cartItems %}
+-           {% set cartItemPrice = cartItemPrices[cartItem.id] %}
+-           {% set cartItemDiscount = cartItemDiscounts[cartItem.id] %}
++       {% for index, cartItem in cartItems %}
++           {% set cartItemPrice = cartItemPrices[index] %}
++           {% set cartItemDiscount = cartItemDiscounts[index] %}
+                <tr class="table-cart__row js-cart-item">
+                    <td class="table-cart__cell table-cart__cell--image">
+```
+
+If you extended `QuantifiedProductPriceCalculation`, check your usage of the protected methods `getTotalPriceWithoutVat()`, `getTotalPriceWithVat()` and `getTotalPriceVatAmount()`.
+The signatures were changed and they now require two parameters to be provided.
+
+This was done to prevent the protected parameters `$quantifiedProduct`, `$product`, and `$productPrice` from being used to pass values between methods.
+These protected parameters were removed.
+Refactor your code if you have used them (you can take a look at [the original commit for inspiration](https://github.com/shopsys/shopsys/pull/821/commits/deb643e3d9d3ee077488250d50cc10936086c6a5)).
```

Comment:
Since instructions are really long, it would be nice to end it with some kind of conclusion...

"The change is huge, we know about it, but we want to keep improving the framework for You. If you find anything we missed or if you'll need to explain anything more, please create an issue or contact us on our slack"

Reply:
A conclusion is a good idea, I've rewritten your suggestion a bit and added links in aa2108474